org-cyf-sdc/content/tools/sprints/3/process_spelunking_workshop.md

```diff
@@ -283,3 +283,273 @@ You can hit "enter" to get it to stop.
+```
+write(2, "Press enter to stop the program."..., 34Press enter to stop the program...) = 34
+```
+So we can see it's the `write` system call.
+
+{{</tabs>}}
+
+Because we haven't pressed "enter" yet, the process is still waiting for us.
+Let's press "enter" to allow the process to end.
+
+When we pressed "enter", the `read` system call was only half written to the screen:
+```
+read(0,
+```
+And now that we pressed "enter", the rest of it got rendered, but it was broken across two lines.
+
+This is because the output of our process and the output of `strace` are getting mixed up.
+This can be really confusing.
+
+> [!NOTE]
+>
+> You might have noticed the `write` *system call* looked a bit odd too:
+> ```
+> write(2, "Press enter to stop the program."..., 34Press enter to stop the program...) = 34
+> ```
+> It shows "Press enter to stop the program" twice.
+>
+> This is also because of the output of the process getting mixed up with  the output of `strace`:
+> `strace` tried to tell us of the `write` *system call*.
+> And while it was telling us, the `write` *system call* happened.
+> Both tried to write "Press enter to stop the program", therefore it appeared twice.
+
+Because all this mixing up is confusing, we'll use two terminals again.
+We'll have to do things a little different this time.
+
+{{<tabs>}}
+===[[Exercise]]===
+On your first terminal, type `strace -o strace.output python3 exercise_3.py`
+You should see the process, but no longer see the *trace*.
+That's because we are sending it to a file: `strace.output`.
+The `-o` flag means "write everything to this file".
+
+On your second terminal, type `less strace.output` to view the *trace*. 
+
+Can you find the `write` system call again?
+Can you see how it is different from the previous time we found it?
+
+===[[Answer]]===
+This time, the system call looks like this:
+```
+write(2, "Press enter to stop the program."..., 34) = 34
+```
+
+It only writes the "Press enter to stop the program" once.
+
+And we get to interact with our program on the other terminal normally.
+
+You can close `less` now (you can close it by pressing `q`).
+{{</tabs>}}
+
+You now know how to get your process traced.
+And you've seen that the `openat` *system call* is used to open the file.
+
+{{<tabs>}}
+===[[Exercise]]===
+On your first terminal, trace the program again, like you did in the previous exercise: `strace -o strace.output python3 exercise_3.py`
+
+On your second terminal, use `lsof` (from earlier) to find the *file descriptor* for `exercise_3_file.txt`.
+Next, find the `openat` system call where `exercise_3_file.txt` is opened.
+
+Can you tell where `strace` tells you what the file descriptor is?
+
+===[[Answer]]===
+I expect you'll see the same numbers as me.
+
+I can see `lsof` telling me that the *file descriptor* is 3.
+
+And the *trace* shows:
+```
+openat(AT_FDCWD, "exercise_3_file.txt", O_RDONLY|O_CLOEXEC) = 3
+```
+
+That last `= 3` is the *file descriptor*.
```

Comment:
I'd maybe want to generalise here also that this is the _return value_ of the syscall? (Maybe comparing with the `34` from the `write` above for "Press enter to stop the program."?)

Reply:
Did mention that it's the return value. But haven't addressed the return value of `write`.
(Do want to do that, just have to figure out where makes the most sense.)